### PR TITLE
Begin to resolve supertype type parameters

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: kotlin-codegen/compiler/src/main/java/com/squareup/moshi/TypeResolver.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  fun resolve(typeName: TypeName): TypeName {
+    return when (typeName) {
+      is ClassName -> typeName
+
+      is ParameterizedTypeName -> {
+        ParameterizedTypeName.get(
+            typeName.rawType, *(typeName.typeArguments.map { resolve(it) }.toTypedArray()))
+      }
+
+      is WildcardTypeName -> {
```
Comment (NightlyNexus) [inline kotlin-codegen/compiler/src/main/java/com/squareup/moshi/TypeResolver.kt]: existing behavior, unrelated to the PR:
i wonder if these types are worth supporting. i'm unsure of the use case.

File: kotlin-codegen/compiler/src/main/java/com/squareup/moshi/AppliedType.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        val argument = mirror.typeArguments[index]
+        return argument.asTypeName()
+      }
+    }
+  }
+
+  override fun toString() = mirror.toString()
+
+  companion object {
+    fun get(typeElement: TypeElement): AppliedType {
```
Comment (JakeWharton) [inline kotlin-codegen/compiler/src/main/java/com/squareup/moshi/AppliedType.kt]: extension? `TypeElement.toAppliedType()`?

File: kotlin-codegen/compiler/src/main/java/com/squareup/moshi/TypeResolver.kt (hunk 1 above)
Comment (swankjesse) [reply kotlin-codegen/compiler/src/main/java/com/squareup/moshi/TypeResolver.kt]: They come up invisibly in Kotlin’s immutable List type.

File: kotlin-codegen/compiler/src/main/java/com/squareup/moshi/AppliedType.kt (hunk 2 above)
Comment (swankjesse) [reply kotlin-codegen/compiler/src/main/java/com/squareup/moshi/AppliedType.kt]: Good call. Will do in follow-up.

**Pull request merged**